src/node_url.cc

```diff
@@ -1351,6 +1351,53 @@ namespace url {
                             v8::NewStringType::kNormal).ToLocalChecked());
   }
 
+  static void ToUSVString(const FunctionCallbackInfo<Value>& args) {
+    Environment* env = Environment::GetCurrent(args);
+    CHECK_GE(args.Length(), 1);
+    Local<String> str;
+    if (!args[0]->ToString(env->context()).ToLocal(&str))
+      return;
+
+    // The algorithm below can be replaced with `Utf8Value`, which
+    // automatically replaces invalid Unicode surrogates with U+FFFD
+    // REPLACEMENT CHARACTER as the Web IDL spec mandates through V8's
+    // `String::REPLACE_INVALID_UTF8`. However, using `Utf8Value` would prevent
+    // us from using the `changed` optimization, and V8 would have to create a
+    // new string even when it is not needed.
+
+    TwoByteValue value(env->isolate(), str);
+    const size_t n = value.length();
+    bool changed = false;
+
+    for (size_t i = 0; i < n; i++) {
+      uint16_t c = value[i];
+      if (c < 0xD800 || c > 0xDFFF) {
+        continue;
+      } else if (0xDC00 <= c || i == n - 1) {
+        changed = true;
+        value[i] = 0xFFFD;
+      } else {
+        uint16_t d = value[i + 1];
+        if (0xDC00 <= d && d <= 0xDFFF) {
+          i++;
+        } else {
+          changed = true;
+          value[i] = 0xFFFD;
+        }
+      }
+    }
```

Comment:
Macros like this might help to make it a bit easier to read in the C++ land...

```c
#define IS_HIGH_SURROGATE(c) (((c) & 0xFC00) == 0xD800)
#define IS_LOW_SURROGATE(c)  (((c) & 0xFC00) == 0xDC00)
```

I think the `i++` part is a bit implicit(if I understand correctly this would make the second branch "unexpected" so "changed = true"). Testing for an unexpected low surrogate first, then for high surrogates checks the following character(or EOL) immediately instead of waiting until the next cycle could be easier to read.